File: packages/engine/src/cpus/6502/instructions/pla.test.js
import { test, expect } from 'vitest'
import { BufferMapper } from '../../../buffer_mapper'
import { CPU } from '..'

test('plp', () => {
  const mmu = new BufferMapper()
  const cpu = new CPU(mmu)
  cpu.sp = 0xff
  cpu.acc = 0x42
  mmu.data = [0x48, 0x68] // PHA PLA
  const g = cpu.run()
  for (let i = 0; i < 3; i++) g.next()
  expect(cpu.sp).toBe(0xfe)
  expect(mmu.data[0x1ff]).toBe(cpu.acc)

  cpu.acc = 0
  for (let i = 0; i < 4; i++) g.next()
  expect(cpu.sp).toBe(0xff)
  expect(cpu.acc).toBe(0x42)
})

File: packages/engine/src/cores/famibob/cpu/instructions/pla.test.zig
const testing = @import("../testing/setup.zig");

test "PLA" {
    var cpu = testing.createCPU();
    cpu.sp = 0xff;
    cpu.acc = 0x42;
    cpu.write(0, 0x48);
    cpu.write(1, 0x68);
    for (0..3) |_| {
        cpu.process();
    }
    try testing.expectEqual(0xfe, cpu.sp);
    try testing.expectEqual(cpu.acc, cpu.read(0x1ff));

    cpu.acc = 0;
    for (0..4) |_| {
        cpu.process();
    }
    try testing.expectEqual(0xff, cpu.sp);
    try testing.expectEqual(0x42, cpu.acc);
}
